ddtrace/tracer/option.go

```diff
@@ -138,6 +138,26 @@ type config struct {
 
 	// enabled reports whether tracing is enabled.
 	enabled bool
+
+	// postProcessor holds the function used to process finished spans of a trace.
+	// It reports whether the trace should be dropped.
+	postProcessor func([]ddtrace.ReadWriteSpan) (dropTrace bool)
+}
+
+// WithPostProcessor sets post processor function f, which allows certain span fields to change
+// after a trace has finished, and can force a trace to be dropped despite any existing sampling
+// rules. Modifying the span can be done by using methods of the ReadWriteSpan interface, such
+// as SetTag. If f returns true, the entire trace will be dropped. If f returns false, then no
+// changes to the sampling method will be made based on the post processor function, though the
+// trace may be dropped in the agent due to other sampling rules.
+//
+// Please note that dropped traces and spans will not be accounted for in APM stats, and that
+// modifying any of the ReadWriteSpans outside the scope of the processor function can and will
+// likely cause a panic.
```

Comment:
Discussed this with @knusbaum and @ajgajg1134.

In order to avoid stats being lost, let's make sure this code does the following when the processor returns true (indicating that the trace should be dropped):
1. If client-side stats are enabled, then don't send the trace to the agent, since the stats will be computed in the tracer.
2. If client side stats are *not* enabled, mark them as "dropped" by the client and send them to the agent anyway. That way the agent can still compute the necessary stats.

Then, you shouldn't have to have this comment at all, since dropping the trace via the processor will not affect stats either way.